### PR TITLE
Add Annotations to CNT  + add code to deprecate stim_channel

### PR DESCRIPTION
**Pull request by massich**
In this PR:
- [x] Fix CNT date parsing
- [x] Add CNT support in `mne.read_annotations`
- [x] Add `stim_channel` parameter to `mne.io.read_raw_cnt` (to later remove stim synthesis)
- [x] Fix tests to use `stim_channel=False`

This needs to go before #6025

### CODE REVIEW

File: mne/io/cnt/_utils.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+from struct import Struct
+from collections import namedtuple
+from math import modf
+from datetime import datetime
+
+from ...utils import warn
+
+
+def _read_TEEG(f, teeg_offset):
```
Comment (agramfort) [inline mne/io/cnt/_utils.py]: TEEG -> teeg in function name

File: mne/io/cnt/cnt.py [hunk 2]
```diff
@@ -1,26 +1,103 @@
...
+
+from struct import unpack, calcsize
+
+from ._utils import _read_TEEG, _get_event_parser, _session_date_2_meas_date
+
+
+def _read_annotations_cnt(fname):
+    """CNT Annotation File Reader.
+
+    This method opens the .cnt, files searches all the metadata to construct
```
Comment (agramfort) [inline mne/io/cnt/cnt.py]: .cnt, files -> .cnt files,

File: mne/io/cnt/cnt.py [hunk 3]
```diff
@@ -1,26 +1,103 @@
...
+    the annotations and parses the event table. Notice that CNT files, can
+    point to a different file containing the events. This case when the
+    event table is separated from the main .cnt IS NOT COVER.
```
Comment (agramfort) [inline mne/io/cnt/cnt.py]: IS NOT COVER ->  is not supported.

File: mne/io/cnt/tests/test_cnt.py [hunk 4]
```diff
@@ -22,9 +26,47 @@ def test_data():
     with pytest.warns(RuntimeWarning, match='number of bytes'):
         raw = _test_raw_reader(read_raw_cnt, montage=None, input_fname=fname,
                                eog='auto', misc=['NA1', 'LEFT_EAR'])
+
+    # make we use annotations event if we synthesized stim
```
Comment (agramfort) [inline mne/io/cnt/tests/test_cnt.py]: make -> make sure

File: mne/io/cnt/_utils.py (hunk 1 above)
Comment (massich) [reply mne/io/cnt/_utils.py]: true, even if its an accronim we have no capitals in any function name.

**Comment by massich** [comment]
@wmvanvliet you added the type 3 events in #3911, can you test that `events_from_annotations` brings you the right onsets?

**Comment by massich** [comment]
for the record, I've checked with files that have cnt and eeglab versions:
- 32 bits is not working (we can open another PR)
- events are ok.
- channel locations seem to be ok. 

It would be good if we can get cnt files with events of type1 and type3 to be added to the test suit (I only have type2).

**Comment by massich** [comment]
TL;DR
=====
Just for the record, testing with some confidential files + some files on the web running the following

```py
import mne
import pytest
import os.path as op
from datetime import datetime
from mne import __file__ as _mne_file
from mne.tests.test_annotations import _assert_annotations_equal

regular_path = op.join(op.dirname(_mne_file), '..', 'sandbox', 'data')
confidential_path = op.join(regular_path, 'confidential', 'cnt')

flankers_path = op.join(regular_path, '914flankers.cnt')

cnt_files_with_eeglab_pair = [
    op.join(confidential_path, 'BoyoAEpic1_16bit.cnt'),
    op.join(confidential_path, 'cont_67chan_resp_32bit.cnt'),
    op.join(confidential_path, 'SampleCNTFile_16bit.cnt')
]


def test_ensure_meas_date(recwarn):
    raw = mne.io.read_raw_cnt(flankers_path, montage=None,
                              date_format='dd/mm/yy')
    meas_date = (datetime
                 .fromtimestamp(raw.info['meas_date'][0])
                 .strftime('%d/%m/%y %H:%M:%S'))

    assert meas_date == '23/09/07 12:22:15'


@pytest.mark.parametrize('fname', cnt_files_with_eeglab_pair, ids=op.basename)
def test_check_cnt_eeglab_pairs(fname, recwarn):
    raw_eeglab = mne.io.read_raw_eeglab(fname.replace('.cnt', '.set'),
                                        montage=None)
    raw_cnt = mne.io.read_raw_cnt(fname, montage=None)

    assert raw_cnt.info.keys() == raw_eeglab.info.keys()
    xx = object_diff(raw_cnt.info, raw_eeglab.info)
    print(xx)
```

I get: 
* The date work properly, 
* the info difference between `cnt` and its transformed `eeglab` version are
  * the `subject_info` anonymization (eeglab is none cnt has empty fields)
  * digitization mismatch (but this is beeing addressed somwhere else)
  * for one file `cnt` has a trigger channel marked as bad that does not appear in the eeglab version of the file. 

```sh
['chs'][63]['cal'] type mismatch (<class 'numpy.ndarray'>, <class 'float'>)
['chs'][63]['coord_frame'] value mismatch (4, 0)
['chs'][63]['loc'] array mismatch
['chs'][63]['unit_mul'] type mismatch (<class 'float'>, <class 'int'>)
['subject_info'] type mismatch (<class 'dict'>, <class 'NoneType'>)
```

**Comment by agramfort** [comment]
ok it's not a blocker the mismatch with eeglab for my point of view.

>

**Comment by agramfort** [comment]
@massich you have a flake8 error https://travis-ci.org/mne-tools/mne-python/jobs/505338464

File: mne/io/fieldtrip/tests/helpers.py [hunk 5]
```diff
@@ -142,26 +144,33 @@ def get_raw_data(system, drop_sti_cnt=True, drop_extra_chs=False):
...
 
     if system == 'CNT':
-        raw_data.drop_channels(['STI 014'])
         events[:, 0] = events[:, 0] + 1
 
-    if isinstance(cfg_local['eventvalue'], np.ndarray):
-        event_id = list(cfg_local['eventvalue'].astype('int'))
-    else:
-        event_id = [int(cfg_local['eventvalue'])]
+    if 'event_id' not in locals():
```
Comment (agramfort) [inline mne/io/fieldtrip/tests/helpers.py]: calling locals() here seems dangerous makes the code harder to read. Can you see how to avoid it? thx

File: mne/io/cnt/tests/test_cnt.py [hunk 6]
```diff
@@ -22,9 +26,47 @@ def test_data():
...
+        raw = read_raw_cnt(input_fname=fname, montage=None, preload=False,
+                           stim_channel=stim_channel)
+    assert ('STI 014' in raw.info['ch_names']) == stim_channel
+
+
+@testing.requires_testing_data
+def test_read_annotations():
+    """Test reading for annotations from a .CNT file."""
+    annot = read_annotations(fname)
+    assert len(annot) == 6
```
Comment (agramfort) [inline mne/io/cnt/tests/test_cnt.py]: I would merge this test with test_compare_events_and_annotations by replacing above the call to _read_annotations_cnt by `read_annotations(fname)`. It will cover the same amount of code and will save time.

File: mne/io/fieldtrip/tests/helpers.py (hunk 5 above)
Comment (agramfort) [reply mne/io/fieldtrip/tests/helpers.py]: forget it it's in tests. Not strongly necessary then but just nice to do.

**Comment by wmvanvliet** [comment]
> @wmvanvliet you added the type 3 events in #3911, can you test that events_from_annotations brings you the right onsets?

@jona-sassenhagen did the actual tests back then. I don't have any CNT files with type 3 events.

**Comment by larsoner** [comment]
@jona-sassenhagen can you test with your event-type-3 CNT file(s)?

File: mne/io/cnt/_utils.py [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+       long Size;       /* Total length of all the events   */
+       long Offset;     /* Hopefully always 0               */
+    } TEEG;
+    """
+    # we use a more descriptive names based on TEEG doc comments
+    Teeg = namedtuple('Teeg', 'event_type total_length offset')
+    teeg_parser = Struct('<Bll')
+
+    f.seek(teeg_offset)
+    return Teeg._make(teeg_parser.unpack(f.read(teeg_parser.size)))
```
Comment (larsoner) [inline mne/io/cnt/_utils.py]: Using private `_make` does not look clean / safe (using private method). Why not just `return Teeg(*teeg_parser.unpack(f.read(teeg_parser.size)))`

File: mne/io/cnt/_utils.py [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+    elif event_type == 3:
+        event_maker = CNTEventType3
+        struct_pattern = '<HBclhhfccc'  # Same as event type 2
+    else:
+        raise ValueError('unknown CNT even type %s' % event_type)
+
+    def parser(buffer):
+        struct = Struct(struct_pattern)
+        for chunk in struct.iter_unpack(buffer):
+            yield event_maker._make(chunk)
```
Comment (larsoner) [inline mne/io/cnt/_utils.py]: same

File: mne/io/cnt/cnt.py [hunk 9]
```diff
@@ -1,26 +1,106 @@
...
+        # 'auto', 'int16', and 'int32'
+        if event_type == CNTEventType3:
+            offset *= n_bytes * n_channels
+        event_time = offset - 900 - (75 * n_channels)
+        event_time //= n_channels * n_bytes
+        return event_time - 1
+
+    with open(fname, 'rb') as fid:
+        fid.seek(SETUP_NCHANNELS_OFFSET)
+        (n_channels,) = unpack('<H', fid.read(calcsize('<H')))
```
Comment (larsoner) [inline mne/io/cnt/cnt.py]: typically (and more compactly) we use and prefer `np.frombuffer` instead of `fid.read` + `unpack`

File: mne/io/cnt/cnt.py [hunk 10]
```diff
@@ -1,26 +1,106 @@
...
+                           duration=duration,
+                           description=description,
+                           orig_time=None)
 
 
 @fill_doc
 def read_raw_cnt(input_fname, montage, eog=(), misc=(), ecg=(), emg=(),
                  data_format='auto', date_format='mm/dd/yy', preload=False,
-                 verbose=None):
+                 stim_channel=True, verbose=None):
```
Comment (larsoner) [inline mne/io/cnt/cnt.py]: Here it should be `None`

File: mne/io/cnt/cnt.py [hunk 11]
```diff
@@ -221,31 +293,39 @@ def _get_cnt_info(input_fname, eog, ecg, emg, misc, data_format, date_format):
...
+                fid.seek(event_offset + 9 + i * event_bytes + 4)
+                offset = np.fromfile(fid, dtype='<i4', count=1)[0]
+                if event_type == 3:
+                    offset *= n_bytes * n_channels
+                event_time = offset - 900 - 75 * n_channels
+                event_time //= n_channels * n_bytes
+                stim_channel[event_time - 1] = event_id
+
+        else:  # when stim_channel_toggle is False
+            pass
```
Comment (larsoner) [inline mne/io/cnt/cnt.py]: I don't see much value in `else: pass` because it's implied by just not having it there; and in this case the comment does not add anything because it's just the conditional checked above

File: mne/io/cnt/cnt.py [hunk 12]
```diff
@@ -272,18 +352,25 @@ def _get_cnt_info(input_fname, eog, ecg, emg, misc, data_format, date_format):
...
+                     'loc': np.zeros(12),
+                     'coil_type': FIFF.FIFFV_COIL_NONE,
+                     'kind': FIFF.FIFFV_STIM_CH}
+        chs.append(chan_info)
+        baselines.append(0)  # For stim channel
+        cnt_info.update(baselines=np.array(baselines), n_samples=n_samples,
+                        stim_channel=stim_channel, n_bytes=n_bytes)
+    else:
+        cnt_info.update(baselines=np.array(baselines), n_samples=n_samples,
+                        n_bytes=n_bytes)
```
Comment (larsoner) [inline mne/io/cnt/cnt.py]: There is redundancy in several conditionals, including here, which could instead be (no need for `else` at all):
```
if stim_channel_toggle:
    ...
    cnt_info.update(stim_channel=stim_channel)
cnt_info.update(baselines=np.array(baselines), n_samples=n_samples,
                n_bytes=n_bytes)
```
Seems simpler (makes it clear no matter what, `cnt_info` is updated with these entries) and is more DRY, no?

File: mne/io/cnt/cnt.py [hunk 13]
```diff
@@ -404,15 +508,30 @@ def _read_segment_file(self, data, idx, fi, start, stop, cals, mult):
...
+
                 block = block[sel, s_offset:n_samps + s_offset]
                 data_[sel] = block
-                data_[-1] = stim_ch[start + sample_start:start + sample_stop]
+                if 'stim_channel' in self._raw_extras[0]:
+                    _data_start = start + sample_start
+                    _data_stop = start + sample_stop
+                    data_[-1] = stim_ch[_data_start:_data_stop]
+                else:
+                    pass
```
Comment (larsoner) [inline mne/io/cnt/cnt.py]: no need for `else: pass`

File: mne/io/cnt/_utils.py (hunk 7 above)
Comment (massich) [reply mne/io/cnt/_utils.py]: 'Cos I did not know how to write it! Thx.

File: mne/io/cnt/cnt.py (hunk 10 above)
Comment (massich) [reply mne/io/cnt/cnt.py]: Shouldn't default to True to not break people's code? At the beginning I just removed the whole stim. But talking to @agramfort it seems that we cannot get read of the stim in cnt until 0.20, unless we backport the addition of this `stim_channel=True` to 0.17.3 and then 0.18 to false/none and remove the stim channel.

File: mne/io/cnt/cnt.py (hunk 12 above)
Comment (massich) [reply mne/io/cnt/cnt.py]: totally.

File: mne/io/cnt/cnt.py (hunk 9 above)
Comment (massich) [reply mne/io/cnt/cnt.py]: if `np.frombuffer` took a file descriptor (or the file was lodaded in a buffer) this could be written like this which is really readable:

 ```py
    with open(fname, 'rb') as fid:
        n_channels = np.frombuffer(
            fid, dtype=np.uint16, offset=SETUP_NCHANNELS_OFFSET)
        sfreq = np.frombuffer(
            fid, dtype=np.uint16, offset=SETUP_RATE_OFFSET)
        event_table_pos = np.frombuffer(
            fid, dtype=np.int32, offset=SETUP_EVENTTABLEPOS_OFFSET)
```

Otherwise `np.fromfile` but you can not get read of the seek.

```py
    with open(fname, 'rb') as fid:
        fid.seek(SETUP_NCHANNELS_OFFSET)
        n_channels = np.fromfile(fid, dtype='<u2', count=1)[0]

        fid.seek(SETUP_RATE_OFFSET)
        sfreq =  np.fromfile(fid, dtype='<u2', count=1)[0]

        fid.seek(SETUP_EVENTTABLEPOS_OFFSET)
        event_table_pos = np.fromfile(fid, dtype='<i4', count=1)[0]
```
But I've no idea which of the 3 is preferable.

Comment (larsoner) [reply mne/io/cnt/cnt.py]: `frombuffer` is what we use in `mne/io/tag.py` and it works with a file descriptor, we changed it to that from `np.fromfile` a year or so ago so I think it's the preferred way

Comment (larsoner) [reply mne/io/cnt/cnt.py]: (or rather I guess we do `frombuffer(fid.read(...), ...)` in `tag.py`)

Comment (larsoner) [reply mne/io/cnt/cnt.py]: I would say do whichever works. Having a `fid.seek` in there is expected/fine

File: mne/io/cnt/cnt.py [hunk 14]
```diff
@@ -1,26 +1,105 @@
...
+    Returns
+    -------
+    onset : array of float, shape (n_annotations,)
+        The starting time of annotations in seconds after ``orig_time``.
+    duration : array of float, shape (n_annotations,)
+        Durations of the annotations in seconds.
+    description : array of str, shape (n_annotations,)
+        Array of strings containing description for each annotation. If a
+        string, all the annotations are given the same description. To reject
+        epochs, use description starting with keyword 'bad'. See example above.
```
Comment (agramfort) [inline mne/io/cnt/cnt.py]: Docstring is wrong. You return an annotations

File: mne/io/cnt/cnt.py [hunk 15]
```diff
@@ -64,15 +143,23 @@ def read_raw_cnt(input_fname, montage, eog=(), misc=(), ecg=(), emg=(),
...
-        and 'dd/mm/yy'.
+    date_format : 'mm/dd/yy' | 'dd/mm/yy'
+        Format of date in the header. Defaults to 'mm/dd/yy'.
     preload : bool | str (default False)
         Preload data into memory for data manipulation and faster indexing.
         If True, the data will be preloaded into memory (fast, requires
         large amount of memory). If preload is a string, preload is the
         file name of a memory-mapped file which is used to store the data
         on the hard drive (slower, requires less memory).
+    stim_channel : bool (default True)
```
Comment (agramfort) [inline mne/io/cnt/cnt.py]: None not true

**Comment by massich** [comment]
@larsoner merge if you are happy. The code synthesizing the stim channel from the events type 3 is still the same, and there's an upcoming PR fixing the overflow + the 32 bits annotations.

File: mne/io/cnt/cnt.py [hunk 16]
```diff
@@ -221,31 +287,36 @@ def _get_cnt_info(input_fname, eog, ecg, emg, misc, data_format, date_format):
...
+                assert teeg_offset == 0  # documentation say this should be 0
+                if event_type == 1:
+                    event_bytes = 8
+                elif event_type in (2, 3):
+                    event_bytes = 19
+                else:
+                    raise IOError('Unexpected event size.')
+
+                # XXX long NumEvents is available, why are not we using it?
+                n_events = event_size // event_bytes
```
Comment (massich) [inline mne/io/cnt/cnt.py]: I guess that #6535 is why.
